### PR TITLE
removed SDI schemas

### PR DESCRIPTION
**Pull request by beliaev-maksim**
Remove sdi schemas from the charm

### CODE REVIEW

**Comment by ca-scribner** [comment]
@beliaev-maksim I'm very curious where you're going with this one :D

**Comment by ca-scribner** [comment]
@beliaev-maksim can you give some background about what this is solving?  afaict it removes SDI, but in a way that makes the code harder to maintain (instead of using SDI, we have a bespoke relation handler in the charm main that mimics SDI).  This feels like a worse situation for the team than using SDI

**Comment by beliaev-maksim** [comment]
@ca-scribner I think we need to start removing SDI and then slowly start moving to pydantic implementation

Obviously, we can rewrite all charms at once and then we will not need to mimic SDI databag. But how realistic is it?

It might be easier to do it iteratively

**Comment by jnsgruk** [comment]
I think we do need to move iteratively, but I agree with @ca-scribner that mimicking the logic isn't a good move.

One of the benefits I perceive from moving away from SDI is the "chattiness" back and forth, and the serialised data structures.

I think the only credible way to do this is to pick one SDI relation at a time, design it carefully with SDI removed and roll out affected charms in `edge`.

This likely means new interface names to avoid disruption, but with the charm-relation-interfaces repo we can at least try to enforce some commonality through prefixed names etc.

In summary: I'm +100 on removing SDI, and thanks @beliaev-maksim for kicking this off - but let's use the chance to just design the thing we want 😎

**Comment by beliaev-maksim** [comment]
@jnsgruk if we go with another interface name
Should we support backwards and keep two interfaces at a time?
Otherwise, during upgrade it might start failing

**Comment by ca-scribner** [comment]
Yeah this interface in particular is a really good one for removing it and replacing with a typical library because it is a sort of "private" interface - the only applications that use this are istio-pilot and oidc-gatekeeper.  We can convert both at the same time and then we both get rid of SDI and the schema SDI imposed.  

If we replace SDI with a lib in place (without changing the relation name/interface), we need to test upgrades.  Even if we say istio/oidc charms only support relating to their comparable version, upgrades will (momentarily) break that as one charm will win the upgrade race and we need to check whether we gracefully recover from that.  Not a hard thing to do, but something that should be specced and the results well documented because it is important to get right (both for this charm, but also because it really matters for all the over SDI conversions, too)

**Comment by ca-scribner** [comment]
I think an easier migration path might be to:
* write a new charm lib for this relation using a simplified (not sdi-like) interface data structure
* write a reusable decorator that offers SDI failover and wrap the standard lib in this

There's design details to sort out, but it could do something like:
* look at the relation to detect if the other side is speaking like an updated relation handler or the standard SDI
  * if the other side is "modern" (not SDI), then delegate everything to the wrapped relation lib
  * if the other side is SDI, convert the results and pass to the wrapped class, as well as notify in logs that this is a deprecated handler and SDI support will be completely removed in the next release

That way for the upcoming release we can write clean relation libraries for all our relations but wrap them in this decorator, then in the following release we just remove the decorator.  Thoughts @jnsgruk @beliaev-maksim ?

**Comment by beliaev-maksim** [comment]
@ca-scribner 
that sounds too much. I am thinking about https://github.com/canonical/charm-relation-interfaces and the core idea of it. Your relation interface must not change under any circumstances. 

while the proposed solution with wrapper will do exactly that.

what we can do instead is enabling second interface as @jnsgruk suggested.

then we have 2 ways: either we keep 2 interfaces on the charm and write some logic to switch between them (like we did for MariaDB and MySQL) or we just go hard way and cut out old interface but keep the relation name. (I assume juju will be able to resolve the issue during migration)

**Comment by ca-scribner** [comment]
@beliaev-maksim yeah I get what you're saying, but I don't know if it is the best practical solution.  Adding a new relation and migrating to the new one might be the right solution, but it also makes the upgrade path awkward (you cannot automated it - there needs to be manual intervention).  Where these interfaces were all designed to be used internally within the Charmed Kubeflow project, I lean toward migrating their schema rather than replacing it entirely.  But I do see your point.

[charm-relation-interfaces](https://github.com/canonical/charm-relation-interfaces) should provide for schema updates - schemas do evolve over time.  I don't see any guidance in the readme.md for this though

**Comment by ca-scribner** [comment]
Thinking more on this, it might just be easier to change the relation/interface names since things are changing so much.  It makes for a painful upgrade, but maybe it is the best of the bad options?  not sure...

**Comment by simskij** [comment]
> @ca-scribner hmm, I think CRI don't support changes in interface to ensure stability. But I might be wrong
> 
> @simskij might suggest more

This is correct. A schema that is established and gone "live" is not to be patched, extended or otherwise changed. Subsequent changes go into a new version of the schema. These versions do not correspond to library version, charm version or anything else.

**Comment by beliaev-maksim** [comment]
@simskij do we have already one example of CRI with multiversion schema ?

**Comment by simskij** [comment]
> @simskij do we have already one example of CRI with multiversion schema ?

Yes, `tls-certificates`. 👍🏼

**Comment by beliaev-maksim** [comment]
@simskij thanks

@ca-scribner 
then we need something similar to https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/tls_certificates
maybe v0 with SDI and v1 without ?
